## Long Range Transportation Plan

In [122]:
import _categorizing_utils as categorizing_utils
import _harmonization_utils as harmonization_utils
import _lrtp_utils as lrtp_utils
import _specific_list_utils as list_utils
import geopandas as gpd
import numpy as np
import pandas as pd
from calitp_data_analysis.sql import to_snakecase

In [123]:
import fsspec

In [124]:
import warnings
warnings.filterwarnings("ignore")

In [125]:
pd.options.display.max_columns = 100
pd.options.display.float_format = "{:.2f}".format
pd.set_option("display.max_rows", None)
pd.set_option("display.max_colwidth", None)

### Ambag

In [126]:
ambag = lrtp_utils.ambag_lrtp()

In [127]:
lrtp_utils.give_info(ambag, "project", "project_description")

Citywide Bike Lanes                         2
The Alameda - Salinas Road Bike Route       2
Citywide Sidewalk Program                   2
Carmel to Pebble Beach Bike/Ped Facility    1
Second Street Bike Lane                     1
Name: project, dtype: int64
# of unique project titles: 277
After dropping duplicates using project and project_description: 280
Df shape: (280, 6)
Index(['ambag_id', 'project', 'project_description', 'total_cost_\n_$_000s_',
       'county', 'total_cost_millions'],
      dtype='object', name=0)


In [128]:
ambag.loc[ambag.project == "Citywide Bike Lanes"]

0         ambag_id              project  \
133  MON-SOL075-SO  Citywide Bike Lanes   
52   SC-SV-P41-SCV  Citywide Bike Lanes   

0                                                                  project_description  \
133               Bike Lanes (2007 TIF M2, 2013 TIF M2); construct bike lanes citywide   
52   Construction of additional bike lanes and paths citywide (including Green Hills).   

0   total_cost_\n_$_000s_      county total_cost_millions  
133                  1440  Santa Cruz             1440000  
52                   2060  San Benito             2060000

### BCAG 

In [129]:
bcag = lrtp_utils.bcag_lrtp()

In [130]:
bcag[['fund_estimate','total_project_cost','rtp_id']].sample(3)

fund_estimate  total_project_cost  rtp_id
63    $1.1 million          1100000.00      61
168  $1.62 million          1620000.00     169
5    $3.42 million          3420000.00       6

In [131]:
lrtp_utils.give_info(bcag, "project_descr", "total_project_cost")

New Traffic Signal                                                                                                                                   6
Turn lane capacity expansion, storage length expansion, channelization improvements, pedestrian safety due to increased traffic volumes.             4
East Gridley Rd. At Feather River, 1.0 mile east of Larkin Rd. Scope is to address cracks with a Methacrylate Deck treatment. Bridge No. 12C0022.    2
Roundabout (within existing ROW)                                                                                                                     2
operational flow improvments (traffic signals or roundabouts)                                                                                        2
Name: project_descr, dtype: int64
# of unique project titles: 239
After dropping duplicates using project_descr and total_project_cost: 247
Df shape: (250, 23)
Index(['rtp_id', 'agency', 'ftip', 'rtp', 'project_type', 'title',
       'project_de

In [132]:
# bcag.loc[bcag.project_descr == "Turn lane capacity expansion, storage length expansion, channelization improvements, pedestrian safety due to increased traffic volumes."]

### Fresno COG

In [133]:
fresno_cog = lrtp_utils.fresnocog_lrtp()

In [134]:
lrtp_utils.give_info(fresno_cog, "project_title", "estimated\ntotal_cost\n_$1,000_")

None                                               3
SR 41 at Elkhorn Ave Roundabout                    2
SR 41 at Mount Whitney Ave Roundabout              2
Ventura, SR 41 to SR 99  Widen from 4LU to 4 LD    2
Herndon-Polk to Weber : 4Ld to 6LD                 2
Name: project_title, dtype: int64
# of unique project titles: 3129
After dropping duplicates using project_title and estimated
total_cost
_$1,000_: 3143
Df shape: (3147, 10)
Index(['agency', 'project_id', 'project_title', 'project_description',
       'project_type', 'estimated_open_to_traffic',
       'estimated\ntotal_cost\n_$1,000_', 'est__o2t_year',
       'est_total_project_cost', 'financial_constraint'],
      dtype='object')


In [135]:
# fresno_cog.loc[fresno_cog.project_type.str.contains("Ac")]

In [136]:
lrtp_utils.group_duplicates(
    fresno_cog,
    "project_title",
    "estimated\ntotal_cost\n_$1,000_",
    "project_description",
)

project_description
project_title                                              estimated\ntotal_cost\n_$1,000_                     
Ventura, SR 41 to SR 99  Widen from 4LU to 4 LD            3427.00                                            2
Herndon-Polk to Weber : 4Ld to 6LD                         2931.00                                            2
10th Ave / Kamm Ave Traffic Signal                         400.00                                             1
Millerton Road From Auberry Rd to SR 168 Shoulder Widening 5910.00                                            1
Midtown Trail: Millbrook - Michigan to McKinley            680.00                                             1
Milburn & Spruce: New Traffic Signal                       620.00                                             1
Mill and Orange Park & Ride Lot                            505.00                                             1
Millbrook & Nees: Left Turn Phasing                        380.00                                             1
Millbrook & Olympic: New Traffic Signal                    500.00                                             1
Millbrook - Alluvial to Shepherd: AC Overlay               656.00                                             1

In [137]:
fresno_cog[fresno_cog.project_title.str.contains("Ventura, SR 41")]

agency project_id                                    project_title  \
989   Fresno  FRE111312  Ventura, SR 41 to SR 99  Widen from 4LU to 4 LD   
2071  Fresno  FRE111312  Ventura, SR 41 to SR 99  Widen from 4LU to 4 LD   

                                                                    project_description  \
989                                                      Widen to 4 LN Divided Arterial   
2071  Widen to 4 LN Divided Arterial(Measure C Project F in the Urban Regional Program)   

                               project_type estimated_open_to_traffic  \
989   Streets & Roads - Capacity Increasing                 2027-2031   
2071    Streets & Roads-Capacity Increasing                      2028   

      estimated\ntotal_cost\n_$1,000_ est__o2t_year  est_total_project_cost  \
989                           3427.00    2028-01-01              3427000.00   
2071                          3427.00           NaT                     NaN   

     financial_constraint  
989           constrained  
2071          constrained

In [138]:
3427*1_000

3427000

### KCAG

In [139]:
kcag = lrtp_utils.kcag_lrtp()

In [140]:
kcag.sample()

category jurisdiction state_route post_mile      location  \
65  STATE HIGHway projects           NaN          41  8.1/16.3  SR 33 to I-5   

             project_limits description title  total_cost  
65  Widen from 2 to 4 lanes         NaN  None           0

In [141]:
# kcag.title.nunique()

In [142]:
lrtp_utils.give_info(kcag, "description", "category")

Overlay                                                                                        33
Widen from 2 to 4 lanes with left turns                                                         4
Reconstruct and improve curb/ramps                                                              3
Intersection Improvements                                                                       2
Overlay of various thicknesses.  Some shoulder work and dig-out and patch prior to overlay.     2
Name: description, dtype: int64
# of unique project titles: 18
After dropping duplicates using description and category: 20
Df shape: (84, 9)
Index(['category', 'jurisdiction', 'state_route', 'post_mile', 'location',
       'project_limits', 'description', 'title', 'total_cost'],
      dtype='object')


In [143]:
lrtp_utils.group_duplicates(kcag, "location", "description", "category")

category
location                    description                                                                                          
10th Avenue                 Widen from 2 to 4 lanes with left turns                                                             1
Purchase CNG Refuse Vehicle Vehicle Purchase                                                                                    1
Lemoore Avenue              Widen to 5 lanes, add 2 lanes with continuous left turn lane                                        1
Linda Lee Circle            Overlay                                                                                             1
Madera Street               Reconstruct and improve curb/ramps                                                                  1
Magnolia Street             Overlay                                                                                             1
Major Roads                 Overlay of various thicknesses.  Some shoulder work and dig-out and patch prior to overlay.         1
Merced Street               Reconstruct and improve curb/ramps                                                                  1
Minor Roads                 Overlay of various thicknesses.  Some shoulder work and dig-out and patch prior to overlay.         1
Montecito Court             Overlay                                                                                             1

In [144]:
kcag.total_cost.describe()

count   84.00
mean     0.00
std      0.00
min      0.00
25%      0.00
50%      0.00
75%      0.00
max      0.00
Name: total_cost, dtype: float64

In [145]:
kcag.sample()

category jurisdiction state_route post_mile  \
33  Proposed Improvements      Lemoore         NaN       NaN   

            location            project_limits description title  total_cost  
33  Linda Lee Circle  Cinnamon Dr. to terminus     Overlay  None           0

### KernCog

In [146]:
kern = lrtp_utils.kern_lrtp()

In [147]:
kern.sample(2)

project_title  \
450  Bakersfield ‐ Class I Shared Use Path   
221   County Areas ‐ active transportation   

                                                            scope  \
450  NE Bakersfield Path (Paladino Dr to Morning Dr Path) 2.7 mi.   
221                Safe complete streets, pedestrian enhancements   

    yoe_w__new_revenue yoe_w_o_new_reven  \
450            2431000           2431000   
221            3225000            112000   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             

In [148]:
lrtp_utils.give_info(kern, "project_title", "cost")

Kern County ‐ Class II Bike Ln        85
Bakersfield ‐ Class III Bike Blvd     85
Bakersfield ‐ Class II Bike Ln        78
Kern County ‐ Sidewalk Improvement    71
Kern County ‐ Class III Bike Blvd     35
Name: project_title, dtype: int64
# of unique project titles: 371
After dropping duplicates using project_title and cost: 1180
Df shape: (1411, 6)
Index(['project_title', 'scope', 'yoe_w__new_revenue', 'yoe_w_o_new_reven',
       'maint__inflation_savings', 'cost'],
      dtype='object')


In [149]:
# kern.loc[kern.project_title == "M St / Belshaw St‐Park St"]

In [150]:
lrtp_utils.group_duplicates(kern, "project_title", "scope", "cost")

cost
project_title                        scope                                                                                           
County Areas ‐ transit               Senior/disabled & advanced technology transit, vanpools, shared ride, aviation                 9
County Areas ‐ active transportation Safe complete streets, pedestrian enhancements                                                 5
County Areas ‐ bike facilities       Construct Class I (trails), II (lanes) or Class III (routes) Bike Paths; striping; signage     3
Kern County ‐ Complete Streets/ITS I Other Future developments funded by a transportation impact fee and mitigation                 3
Kern County ‐ Corridor Improvement   Hall Rd (San Emidio St to Habecker Rd) .5 mi.                                                  2
Route 58                             2Future Rt 58 from I‐5 to Heath Rd at Stockdale Hwy ‐ construct new freeway                    2
Kern County ‐ Corridor Improvement   Myrtle Av (Panama Ln to Wharton Av) .9 mi.                                                     2
Caltrans ‐ Class II Bike Ln          E Bear Mtn Blvd SR 223 (Comanche Dr to Union St) 12.6 mi.                                      2
Santa Fe Way                         1Hageman to Los Angeles Ave ‐ widen to four lanes                                              2
Kern County ‐ Corridor Improvement   San Diego St (Burgundy Av to Wharton Av) .7 mi.                                                2

In [151]:
kern[kern.project_title == "US 395"].sort_values(["cost"])

project_title  \
155        US 395   
161        US 395   

                                                            scope  \
155  Between Rt 178 and China Lake Blvd ‐ construct passing lanes   
161            San Bdo County Line to Rt 14 ‐ widen to four lanes   

    yoe_w__new_revenue yoe_w_o_new_reven maint__inflation_savings       cost  
155           15117000         191000000                  4883000   15117000  
161          184422000         451035000                 59578000  184422000

In [152]:
kern[kern.project_title == "Teh. Willow Springs Rd"].sort_values(["cost"])

project_title                                         scope  \
1115  Teh. Willow Springs Rd  Rt 58 to Rosamond Blvd ‐ widen to four lanes   
1322  Teh. Willow Springs Rd  Rt 58 to Rosamond Blvd ‐ widen to four lanes   

     yoe_w__new_revenue yoe_w_o_new_reven maint__inflation_savings       cost  
1115          114101000         150961000                 36861000  114101000  
1322          377914000         377914000                122086000  377914000

### Madera CTC

In [153]:
madera = lrtp_utils.madera_lrtp()

In [154]:
lrtp_utils.group_duplicates(madera, "project_name", "total_cost", "description")

description
project_name                                                 total_cost                 
Purchase Transit Vehicle                                     300000                    3
Road 36                                                      563000                    3
Avenue 12                                                    10000000                  3
Various                                                      1000000                   3
Sunset Avenue & N Westberry Blvd                             $                         2
Pecan Avenue                                                 665000                    2
Avenue 9                                                     567000                    2
James Madison Elementary School SRTS Pedestrian Improvements $                         2
Northwest Downtown Pedestrian Access                         $                         2
SR 41                                                        $            -            2

In [155]:
lrtp_utils.give_info(madera, "project_name", "total_cost")

Various                            20
Avenue 12                          10
Avenue 9                            8
State Route 41                      6
Purchase Transit Van (Electric)     5
Name: project_name, dtype: int64
# of unique project titles: 515
After dropping duplicates using project_name and total_cost: 735
Df shape: (765, 8)
Index(['category', 'agency', 'project_name', 'location', 'project_id',
       'description', 'opening_year', 'total_cost'],
      dtype='object')


In [156]:
madera[madera.project_name == "Purchase Transit Vehicle"]

category          agency              project_name location  \
2   Public Transit  City of Madera  Purchase Transit Vehicle      NaN   
3   Public Transit  City of Madera  Purchase Transit Vehicle      NaN   
4   Public Transit  City of Madera  Purchase Transit Vehicle      NaN   
23  Public Transit   Madera County  Purchase Transit Vehicle      NaN   
24  Public Transit   Madera County  Purchase Transit Vehicle      NaN   

   project_id                   description  opening_year total_cost  
2   MAD213201  Purchase new transit vehicle           NaN     300000  
3   MAD213202  Purchase new transit vehicle           NaN     300000  
4   MAD213203  Purchase new transit vehicle           NaN     300000  
23  MAD115006  Purchase new transit vehicle           NaN     825000  
24  MAD115010  Purchase new transit vehicle           NaN      57000

In [157]:
madera[madera.project_name == "Avenue 12"].head()

category         agency project_name                    location  \
46  Streets and Roads  Madera County    Avenue 12          Road 38 to Road 40   
47  Streets and Roads  Madera County    Avenue 12  Road 40 to Riverwalk\nBlvd   
48  Streets and Roads  Madera County    Avenue 12    Riverwalk Blvd to SR\n41   
49  Streets and Roads  Madera County    Avenue 12        SR 41 To Flagbarn Rd   
51  Streets and Roads  Madera County    Avenue 12      Road 30 1/2 to Road 36   

   project_id         description  opening_year total_cost  
46        NaN  2 Lanes to 4 Lanes       2027.00   10000000  
47        NaN  2 Lanes to 6 Lanes       2027.00   10000000  
48        NaN  4 Lanes to 8 Lanes       2027.00   10000000  
49        NaN  2 Lanes to 4 Lanes       2030.00    4250000  
51        NaN  2 Lanes to 4 Lanes       2030.00   21000000

### MCAGOV

In [158]:
mcagov = lrtp_utils.mcagov_lrtp()

In [159]:
lrtp_utils.give_info(mcagov, "title", "total_cost\n_$1,000s_")

Bellevue Rd Realignment/Reconstruction    1
Minturn Road Rehabilitation               1
MCAG Vanpool Program (dibs)               1
Measure V Transit Free Ride Program       1
Planada Route Improvements                1
Name: title, dtype: int64
# of unique project titles: 108
After dropping duplicates using title and total_cost
_$1,000s_: 108
Df shape: (108, 8)
Index(['agency', 'title', 'limits_description', 'type', 'completion\nyear',
       'total_cost\n_$1,000s_', 'funding_sources', 'total_cost_millions'],
      dtype='object')


In [160]:
mcagov.sample()

agency                           title  \
12  Gustine  Gustine Phase 3 Multi-use Path   

                                                             limits_description  \
12  Extension of multi-use path on Meredith north on Railroad Ave CML-5230(015)   

                 type completion\nyear total_cost\n_$1,000s_ funding_sources  \
12  Active (Bike/Ped)             2025                   638            CMAQ   

   total_cost_millions  
12              638000

### MTC
* 133 projects but point and line map returns 240 ish rows.
* The geospatial projects are split into various phases while the regular pandas df is just the project in its entirety.
* https://www.planbayarea.org/2050-plan/final-plan-bay-area-2050/final-supplemental-reports/interactive-transportation-project-list
* https://opendata.mtc.ca.gov/datasets/MTC::2023-transportation-improvement-program-projects-point/about
* https://opendata.mtc.ca.gov/datasets/MTC::plan-bay-area-2050-transportation-projects-line/about
* https://data.bayareametro.gov/Projects/Plan-Bay-Area-2050-Transportation-Project-List/y7ka-jksz

In [161]:
mtc_og = lrtp_utils.mtc_lrtp()

In [162]:
# mtc_og.drop(columns = ['geometry','scope']).sample()

In [163]:
#layer_list = ["planbayarea2050_rtp_line", "planbayarea2050_rtp_point"]
#url_pt_1 = "https://services3.arcgis.com/i2dkYWmb4wHvYPda/ArcGIS/rest/services/"
#url_pt_2 = "/FeatureServer/0/query?where=1%3D1&objectIds=&time=&geometry=&geometryType=esriGeometryEnvelope&inSR=&spatialRel=esriSpatialRelIntersects&resultType=none&distance=0.0&units=esriSRUnit_Meter&relationParam=&returnGeodetic=false&outFields=*+&returnGeometry=true&featureEncoding=esriDefault&multipatchOption=xyFootprint&maxAllowableOffset=&geometryPrecision=&outSR=&defaultSR=&datumTransformation=&applyVCSProjection=false&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnExtentOnly=false&returnQueryGeometry=false&returnDistinctValues=false&cacheHint=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&having=&resultOffset=&resultRecordCount=&returnZ=false&returnM=false&returnExceededLimitFeatures=true&quantizationParameters=&sqlFormat=none&f=pgeojson&token="
#gdf = lrtp_utils.open_rest_server(url_pt_1, url_pt_2, layer_list)

In [164]:
#gdf.columns

In [165]:
#gdf.loc[gdf.rtp_id == "21-T10-088"]

In [166]:
#all_projects_url = "https://data.bayareametro.gov/resource/y7ka-jksz.json?$query=SELECT%0A%20%20%60plan_strategy%60%2C%0A%20%20%60rtpid%60%2C%0A%20%20%60title%60%2C%0A%20%20%60scope%60%2C%0A%20%20%60open_period%60%2C%0A%20%20%60funding_millions_yoe%60%2C%0A%20%20%60county%60"
#all_projects = pd.read_json(all_projects_url)

In [167]:
#all_projects.loc[all_projects.rtpid == "21-T10-088"]

In [168]:
#mtc_og.loc[mtc_og.rtpid == "21-T10-088"].drop(columns = ['scope','geometry'])

In [169]:
#mtc_og.loc[mtc_og.proj_title == "Light Rail | Service Expansion | VTA | Stevens Creek Blvd"].drop(columns = ['scope','geometry'])

### Sacramento Area Council of Governments (SACOG)
* https://www.sacog.org/sites/main/files/file-attachments/2020_mtp-scs.pdf?1580330993

In [170]:
sacog = lrtp_utils.sacog_lrtp()

96 rows are headers


In [171]:
lrtp_utils.give_info(sacog, "title", "description")

Safety Improvements                            2
Connect Card Implementation                    2
Aux Lane Project: EB Latrobe Road              1
Bus Maintenance Facility #1 Rehabilitation     1
Enhanced Employee Protection Warning System    1
Name: title, dtype: int64
# of unique project titles: 1599
After dropping duplicates using title and description: 1601
Df shape: (1601, 10)
Index(['id', 'status__planned,_programmed_or_project_development_only_',
       'county', 'lead_agency', 'budget_category', 'title', 'description',
       'total_project_cost__2018_dollars_',
       'year_of_expenditure_cost_for_planned_projects', 'completion_timing'],
      dtype='object')


In [172]:
sacog.sample()

id status__planned,_programmed_or_project_development_only_ county  \
396  PLA25536                                               Programmed    PLA   

       lead_agency                  budget_category                     title  \
396  Placer County  C- Maintenance & Rehabilitation  Crosby Harold Rd. Bridge   

                                                                                                                                                                                description  \
396  Crosby Harold Rd. Over Doty Creek, 0.9 mi N of Wise Rd.: Replace an existing 1 lane bridge with a new 2 lane bridge. (Toll Credits for PE, ROW, CON).  Toll\nCredits for ENG, ROW, CON   

    total_project_cost__2018_dollars_  \
396                           5000000   

    year_of_expenditure_cost_for_planned_projects completion_timing  
396                                           NaN         2020-2025

### SANDAG
* https://www.arcgis.com/home/item.html?id=ba8b2a6247ef463cb7411f039959f3ee
* https://www.arcgis.com/apps/mapviewer/index.html?webmap=ba8b2a6247ef463cb7411f039959f3ee

In [173]:
sandag = lrtp_utils.sandag_lrtp()

In [174]:
lrtp_utils.group_duplicates(sandag, "project_name", "description", "total_project_cost")

total_project_cost
project_name                   description                                                                                         
LRT 530                        Green Line (Santee to Downtown; Double/Third-tracking and Grade Separations)                       2
Rapid 630                      Iris Trolley/Palomar to Kearny Mesa via I-5/163 and City College                                   1
I-5 (Cassidy St to Harbor Dr.) 6 Freeway Lanes and 4 Managed Lanes                                                                1
Rapid 471                      Downtown Escondido to East Escondido                                                               1
Rapid 450 Phase 2              Oceanside to Escondido via Palomar Airport Rd and SR 78 (Full version of Rapid)                    1
Rapid 450 Phase 1              Oceanside to Escondido via Palomar Airport Rd and SR 78 (Light version of Rapid)                   1
Rapid 440                      Carlsbad to Escondido Transit Center via Palomar Airport Rd                                        1
Rapid 41                       Fashion Valley to UTC/UC San Diego via Linda Vista and Clairemont                                  1
Rapid 30                       Balboa Station to Sorrento Mesa via Pacific Beach, La Jolla, UTC                                   1
Rapid 295                      Spring Valley to Clairemont via La Mesa & Kearny Mesa                                              1

In [175]:
sandag.loc[sandag.project_name == "LRT 530"][
    ["project_name", "total_project_cost", "description"]
]

project_name  total_project_cost  \
159      LRT 530        192000000.00   
160      LRT 530        192000000.00   

                                                                      description  
159  Green Line (Santee to Downtown; Double/Third-tracking and Grade Separations)  
160  Green Line (Santee to Downtown; Double/Third-tracking and Grade Separations)

### SBCAG

In [176]:
sbcag = lrtp_utils.sbcag_lrtp()

65 rows are headers


In [177]:
lrtp_utils.give_info(sbcag, "project_title", "total_cost__$000s_")

SB-PL-15: Upper De la Vina St Gap Closure and Safe Crossings              2
CT-1: SR 246 Passing Lanes – Planting Mitigation (FTIP CT93)(EA 0C641)    1
SM-PL-10: A Street Widening                                               1
Sol-PL-6: Fredensborg Canyon Rd/Adobe Creek Culvert Replacement           1
Sol-PL-5: Solvang School Sidewalk Project                                 1
Name: project_title, dtype: int64
# of unique project titles: 418
After dropping duplicates using project_title and total_cost__$000s_: 419
Df shape: (419, 8)
Index(['type', 'project_title', 'phase', 'description',
       'primary_funding_source_s_', 'year', 'total_cost__$000s_',
       'total_cost_millions'],
      dtype='object')


In [178]:
sbcag.sample()

type                                         project_title    phase  \
210  Planned  Gu-PL-1: Various Transportation Improvement Projects  VARIOUS   

                                                                                          description  \
210  Utilize projected Measure A funds to implement various local transportation improvement projects   

    primary_funding_source_s_         year total_cost__$000s_  \
210                       NaN  2026-\n2050              16309   

    total_cost_millions  
210            16309000

### Shasta County Regional Transportation Planning Agency (SCRTPA)

In [179]:
scrtpa = lrtp_utils.scrtpa_lrtp()

In [180]:
lrtp_utils.give_info(scrtpa, "regional_transportation_projects", "cost")

E Cypress Ave-Alfreda Way-Victor Ave                      2
GROVE ST-B ST-WALNUT ST                                   2
Churn Creek Rd-Rancho Rd-Churn Creek Rd/S Bonnyview Rd    2
Hartnell Ave-Victor Ave-Shasta View Dr                    2
Hawley Rd-Hawley Rd (e/o Norwich Ct)-Collyer Dr           2
Name: regional_transportation_projects, dtype: int64
# of unique project titles: 1051
After dropping duplicates using regional_transportation_projects and cost: 1066
Df shape: (1066, 7)
Index(['regional_transportation_projects',
       'short_term_total_est_cost_of_project',
       'long_term_total_est_cost_of_project', 'project_band',
       'project_type\n_project_intent_', 'expected_funding_sources', 'cost'],
      dtype='object')


In [181]:
scrtpa[
    scrtpa.regional_transportation_projects == "E Cypress Ave-Alfreda Way-Victor Ave"
]

regional_transportation_projects  \
831  E Cypress Ave-Alfreda Way-Victor Ave   
833  E Cypress Ave-Alfreda Way-Victor Ave   

     short_term_total_est_cost_of_project long_term_total_est_cost_of_project  \
831                                   NaN                               29913   
833                                   NaN                              529884   

    project_band project_type\n_project_intent_ expected_funding_sources  \
831  (2026-2040)             Buffered Bike Lane                  unknown   
833  (2026-2040)                       Sidewalk                  unknown   

       cost  
831   29913  
833  529884

### San Luis Obispo Council of Governments (SLOCOG)

In [182]:
slocog = lrtp_utils.slocog_lrtp()

In [183]:
lrtp_utils.give_info(slocog, "project_name", "total_cost")

Hazel Ln. SRTS Improvements                          2
Main St. Sidewalks                                   2
Los Berros Rd. Widening                              2
Octagon Barn Park-and-Ride Lot and shared-use lot    1
Augusta Neighborhood SRTS Improvements               1
Name: project_name, dtype: int64
# of unique project titles: 416
After dropping duplicates using project_name and total_cost: 420
Df shape: (420, 12)
Index(['index', 'project_name', 'project_description', 'sponsor', 'community',
       'time_period', 'project_type', '_2023_rtp_project_id', 'time_horizon',
       'total_cost', 'project_id', 'geometry'],
      dtype='object')


In [184]:
# slocog[slocog.project_name == "Hazel Ln. SRTS Improvements"].sample()

### Southern California Association of Governments (SCAG) 

In [185]:
scag = lrtp_utils.scag_lrtp()

360 rows are headers


In [186]:
lrtp_utils.give_info(scag, "description", "route_name")

WIDEN FROM 2 TO 4 LANES      170
WIDEN FROM 4 TO 6 LANES       64
WIDEN FROM 2 TO 6 LANES       43
CONSTRUCT 4 LANE ARTERIAL     21
WIDEN FROM 3 TO 4 LANES       10
Name: description, dtype: int64
# of unique project titles: 2607
After dropping duplicates using description and route_name: 2824
Df shape: (2952, 13)
Index(['system', 'lead_agency', 'rtp_id', 'route_#', 'route_name', 'from',
       'to', 'description', 'completion_year', 'project_cost__$1,000s_',
       'county', 'project_title', 'project_cost_millions'],
      dtype='object')


In [187]:
lrtp_utils.group_duplicates(
    scag, "description", "project_cost__$1,000s_", "project_title"
)

project_title
description             project_cost__$1,000s_               
WIDEN FROM 2 TO 4 LANES 3000                                5
                        6000                                3
                        3210                                2
                        1878                                2
WIDEN FROM 4 TO 6 LANES 10000                               2
WIDEN FROM 2 TO 4 LANES 6901                                2
                        8000                                2
WIDEN FROM 4 TO 6 LANES 5000                                2
                        20000                               2
WIDEN FROM 2 TO 4 LANES 9571                                2

In [188]:
scag.loc[
    (scag.description == "WIDEN FROM 4 TO 6 LANES")
    & (scag["project_cost__$1,000s_"] == 20000)
]

system          lead_agency     rtp_id route_#      route_name  \
1441  LOCAL HIGHWAY               PERRIS  3A01WT105       0     PERRIS BLVD   
1595  LOCAL HIGHWAY  RIVERSIDE,  CITY OF    3A07315       0  VAN BUREN BLVD   

              from        to              description completion_year  \
1441  RAMONA EXPWY  NUEVO RD  WIDEN FROM 4 TO 6 LANES            2025   
1595    AUDREY AVE  GARFIELD  WIDEN FROM 4 TO 6 LANES            2026   

     project_cost__$1,000s_ county project_title project_cost_millions  
1441                  20000    NaN      No Title              20000000  
1595                  20000    NaN      No Title              20000000

###  SJCOG

In [189]:
sjcog = lrtp_utils.sjcog_lrtp()

In [190]:
lrtp_utils.give_info(sjcog, "project_description", "total")

Widen from 2 to 6 lanes                        9
Install ramp meters and ITS elements           7
Rehabilitation of various streets and roads    7
Widen from 2 to 4 lanes                        6
Reconstruct interchange                        5
Name: project_description, dtype: int64
# of unique project titles: 208
After dropping duplicates using project_description and total: 252
Df shape: (262, 16)
Index(['_2018_rtp_mpo_id', 'ctips_id_#', 'jurisdiction', 'facility_name_route',
       'project_description', 'project_limits', 'total', 'ftip_programming',
       'nepa_approval', 'open_to_traffic', 'completion', 'completion_date',
       'facility_name_route_1', 'project_description_1', 'milestone_years_1',
       'project_name'],
      dtype='object')


In [191]:
lrtp_utils.group_duplicates(sjcog, "project_description", "total", "project_name")

project_name
project_description                                                                                                                                                  total                     
Widen from 6 to 8 lanes (inside/outside)                                                                                                                             90000000.00              3
Modification of existing interchange - ENVIRONMENTAL ONLY                                                                                                            2500000.00               2
Reconstruct interchange                                                                                                                                              30000000.00              2
Install ramp meters and ITS elements                                                                                                                                 23400000.00              2
Widen from 2 to 6 lanes                                                                                                                                              9000000.00               2
Widen 6 to 8 lanes (inside)                                                                                                                                          90000000.00              2
Rehabilitation of various streets and roads                                                                                                                          197236934.00             1
Reconstruct interchange to provide 6 through lanes on SR 99, 4 lanes on Harney between Reynolds Ranch Pkwy and SR 99 and modify on-ramps and off-ramps               35000000.00              1
Reconstruct interchange to provide operational and safety improvements on SR 99 at Turner Road (PM 31.3/31.6)                                                        6331338.00               1
Reconstruct intersection, including addition of turn pockets, improvement of traffic signal and installation of train pre-emption system for UPRR railroad crossing. 1151725.06               1

In [192]:
sjcog.loc[
    sjcog.project_description == "Widen from 6 to 8 lanes (inside/outside)"
].head()

_2018_rtp_mpo_id ctips_id_# jurisdiction  facility_name_route  \
1        SJ07-1003        NaN     Caltrans  I-205 Managed Lanes   
2        SJ14-1001        NaN     Caltrans  I-205 Managed Lanes   
3        SJ14-1002        NaN     Caltrans  I-205 Managed Lanes   

                        project_description  \
1  Widen from 6 to 8 lanes (inside/outside)   
2  Widen from 6 to 8 lanes (inside/outside)   
3  Widen from 6 to 8 lanes (inside/outside)   

                           project_limits       total ftip_programming  \
1  Alameda County Line to Eleventh Street 90000000.00              NaN   
2      Eleventh Street to MacArthur Drive 90000000.00              NaN   
3                  MacArthur Drive to I-5 90000000.00              NaN   

  nepa_approval open_to_traffic  completion  completion_date  \
1       2022.00         2028.00         NaN              NaN   
2       2022.00         2028.00         NaN              NaN   
3       2022.00         2028.00         NaN              NaN   

  facility_name_route_1 project_description_1  milestone_years_1 project_name  
1                   NaN                   NaN                NaN         None  
2                   NaN                   NaN                NaN         None  
3                   NaN                   NaN                NaN         None

In [193]:
sjcog[~sjcog.completion_date.isna()].sample()

_2018_rtp_mpo_id ctips_id_# jurisdiction facility_name_route  \
166        SJ07-6021        NaN        SJRRC      ACE Operations   

                                                                                                                                           project_description  \
166  ACE operations and Capital Access Fee (5 trains from 2012 to 2016, 6 trains from 2017 to 2021, 7 trains from 2022 to 2029 and 8 trains from 2030 to 2041)   

                                    project_limits        total  \
166  SJRRC/Santa Clara/Alameda contributions shown 536397757.00   

    ftip_programming nepa_approval open_to_traffic  completion  \
166              NaN           NaN             NaN         NaN   

     completion_date facility_name_route_1 project_description_1  \
166          2030.00                   NaN                   NaN   

     milestone_years_1 project_name  
166                NaN         None

### Stancog

In [194]:
stancog = lrtp_utils.stancog_lrtp()

68 rows are headers


In [195]:
stancog.columns

Index(['jurisdiction', 'location', 'project_limits', 'description',
       'total_cost', 'open_to_traffic', 'funding_source', 'system_preserv_',
       'capacity_enhance_', 'safety', 'oper_', 'complete_streets',
       'active\ntransporta_tion', 'transit', 'other', 'title'],
      dtype='object', name=0)

In [196]:
lrtp_utils.give_info(stancog, "description", "total_cost")

Intersection Improvements               19
Install Traffic Signal                  15
Widen from 2 to 4 lanes                  8
Shoulder Widening - Class 2 Bikepath     7
Widen to 3 lanes                         7
Name: description, dtype: int64
# of unique project titles: 417
After dropping duplicates using description and total_cost: 519
Df shape: (552, 16)
Index(['jurisdiction', 'location', 'project_limits', 'description',
       'total_cost', 'open_to_traffic', 'funding_source', 'system_preserv_',
       'capacity_enhance_', 'safety', 'oper_', 'complete_streets',
       'active\ntransporta_tion', 'transit', 'other', 'title'],
      dtype='object', name=0)


In [197]:
lrtp_utils.group_duplicates(stancog, "project_limits", "description", "total_cost")

0                                                                                                                                                                 total_cost
project_limits                                    description                                                                                                               
Intersection                                      Install Signal & Intersection Improvements                                                                               3
Main Street to T Street                           Pavement Rehabilitation, Maintenance                                                                                     3
Various Locations                                 Pedestrian Improvements                                                                                                  2
Driskell to Sherman Parkway                       Pavement Rehabilitation, Maintenance, and complete streets improvements                                                  2
Highway 33 to Canal School Rd                     Install Collector Street improvements                                                                                    2
Yolo St to Jensen Rd                              Bicycle and Pedestrian Improvements                                                                                      2
SR-108 at First Street                            Install Congestion Management improvements                                                                               1
SR-108 at Claus                                   Install signal light at Claus & SR-108 and Install congestion Management Improvements at First Street & SR-108           1
SR-108 to Riverbank City Limits (Near Oakdale Rd) Multi-Use Path (Class 1) and pedestrian improvements.  (Non-Motorized Transportation Plan Top 25: Route 22)              1
S St to Q St                                      Class 2 Bicycle Lane. Lane narrowing. StanCOG Non-Motorized Transportation Master Plan Project ID: NEW-2                 1

In [198]:
stancog.loc[(stancog.description == "Install Traffic Signal")
]

0  jurisdiction          location                 project_limits  \
0         Ceres         Morgan Rd         Service Rd & Morgan Rd   
4         Ceres  Crows Landing Rd    Crows Landing Rd & A Street   
6         Ceres          Hatch Rd       Hatch Rd & Faith Home Rd   
9         Ceres  Crows Landing Rd  Crows Landing Rd & Grayson Rd   
10        Ceres      Service Road          Service Road & Ustick   
11        Ceres      Whitmore Ave      Whitmore Ave. @  E Street   
12        Ceres      Whitmore Ave       Whitmore Ave & Boothe Rd   
13        Ceres      Whitmore Ave        Whitmore Ave. @ Knox Rd   
16        Ceres  Crows Landing Rd    Crows Landing Rd & B Street   
17        Ceres         Ustick Rd           Ustick Rd & F Street   
18        Ceres      Whitmore Ave    Whitmore Ave. and Ustick Rd   
20        Ceres         Ustick Rd           Ustick Rd & C Street   
21        Ceres      Whitmore Ave   Whitmore Ave & Faith Home Rd   
22        Ceres         Ustick Rd           Ustick Rd & G Street   
81       Newman     SR-33 (South)     Highway 33/Sherman Parkway   

0              description total_cost open_to_traffic  \
0   Install Traffic Signal     347800            2022   
4   Install Traffic Signal     430500            2025   
6   Install Traffic Signal     600000            2025   
9   Install Traffic Signal     499100            2040   
10  Install Traffic Signal     499100            2035   
11  Install Traffic Signal     499100            2030   
12  Install Traffic Signal     514000            2025   
13  Install Traffic Signal     545300            2030   
16  Install Traffic Signal     578500            2030   
17  Install Traffic Signal     578500            2030   
18  Install Traffic Signal     578500            2030   
20  Install Traffic Signal     670700            2030   
21  Install Traffic Signal     670700            2035   
22  Install Traffic Signal     777500            2040   
81  Install Traffic Signal    1900200            2025   

0                 funding_source system_preserv_ capacity_enhance_ safety  \
0                      PFF, CMAQ             NaN               NaN      X   
4                      WLSP, PFF             NaN               NaN      X   
6                      CMAQ, PFF             NaN               NaN      X   
9                      CMAQ, PFF             NaN               NaN      X   
10                     WLSP, PFF             NaN               NaN      X   
11                     WLSP, PFF             NaN               NaN      X   
12                     CMAQ, PFF             NaN               NaN      X   
13                     WLSP, PFF             NaN               NaN      X   
16                     WLSP, PFF             NaN               NaN      X   
17                     WLSP, PFF             NaN               NaN      X   
18                     WLSP, PFF             NaN               NaN      X   
20                     WLSP, PFF             NaN               NaN      X   
21                     CMAQ, PFF             NaN               NaN      X   
22                     WLSP, PFF             NaN               NaN      X   
81  CFF, LTF, CMAQ, STBGP, Local             NaN               NaN      X   

0  oper_ complete_streets active\ntransporta_tion transit other title  
0      X              NaN                     NaN     NaN   NaN  None  
4      X              NaN                     NaN     NaN   NaN  None  
6      X              NaN                     NaN     NaN   NaN  None  
9      X              NaN                     NaN     NaN   NaN  None  
10     X              NaN                     NaN     NaN   NaN  None  
11     X              NaN                     NaN     NaN   NaN  None  
12     X              NaN                     NaN     NaN   NaN  None  
13     X              NaN                     NaN     NaN   NaN  None  
16     X              NaN                     NaN     NaN   NaN  None  
17     X              NaN                    

### TMPO (Tahoe)
* Open data portal: https://www.tahoeopendata.org/search?tags=transportation
* https://www.tahoeopendata.org/datasets/TRPA::2020-regional-transportation-plan-projects/about

In [199]:
tahoe = lrtp_utils.tmpo_lrtp()

### TCAG Tulare County Association of Governments (TCAG)

In [200]:
tcag = lrtp_utils.tcag_lrtp()

In [201]:
tcag.groupby(["project_title", "cost_x_1,000"]).agg(
    {"project_category": "count"}
).sort_values("project_category", ascending=False).head()

project_category
project_title                                             cost_x_1,000                  
Bike/Ped Improvements next 9 years                        0.00                         6
Bike/Ped Improvements next 9 yrs second half of measure R 500.00                       3
Avenue 316 Construct new roadway                          6500.00                      3
Transit Center Minor                                      300.00                       2
Bike/Ped Improvements next 9 years                        390.00                       2

In [202]:
lrtp_utils.group_duplicates(
    tcag, "project_title", "project_description", "cost_x_1,000"
)

cost_x_1,000
project_title                                             project_description                                                                           
Bike/Ped                                                  Various                                                                                      2
Santa Fe Trail Regional                                   Santa Fe Trail Regional                                                                      2
Purchase Buses                                            Fleet replacement, expansion and electrification, 2036-2046                                  2
                                                          Fleet replacement, expansion and electrification, 2022-2035                                  2
Bike/Ped Improvements next 9 yrs second half of measure R used $49,000 for alta heritage; $50K used for El Monte Way,                                  2
Bike/Ped Improvements next 9 years                        Used $100,000 and $200,000 to Alta Heritage and Roosevelt Paseo respectively                 2
SR 198 Construct new I/C                                  SR-198 at Road 148 (Tower) Construct new interchange                                         1
SR 198 Intersection Improvements                          SR-198 at Spruce Rd Roundabout and local street improvements                                 1
SR 198 Operational  I/C improve.                          SR-198 at Lovers Lane Intersection, rehab, operational improvements                          1
                                                          SR-198 downtown corridor interchanges Turn lane, intersection, ramp improvements             1

In [203]:
tcag.loc[
    (tcag.project_title == "Santa Fe Trail Regional")
    & (tcag.project_description == "Santa Fe Trail Regional")
]

project_id__if_applicable_         jurisdiction            project_title  \
49                        NaN   Visalia (Regional)  Santa Fe Trail Regional   
50                        NaN                 TCAG  Santa Fe Trail Regional   

        project_description       project_category completed_by_2035_y_n  \
49  Santa Fe Trail Regional  Active Transportation                     Y   
50  Santa Fe Trail Regional  Active Transportation                     Y   

   local regional state federal  cost_x_1,000 completion_year  \
49   NaN        X   NaN     NaN        586.37            2035   
50   NaN        X   NaN     NaN       2413.63            2035   

    cost_in_millions  
49         586370.00  
50        2413630.00

## Stack

### Add Lost

In [204]:
all_mpo_df, all_mpo_gdf = lrtp_utils.all_mpo(True)

96 rows are headers
65 rows are headers
360 rows are headers
68 rows are headers


In [205]:
# los_df = harmonization_utils.load_lost()

In [206]:
# los_df.sample()

In [207]:
len(all_mpo_df)

16276

In [208]:
all_mpo_gdf.shape

(1355, 10)

In [209]:
all_mpo_gdf.data_source.unique()

array(['Bcag Lrtp', 'Mtc Lrtp', 'Sandag Lrtp', 'Slocog Lrtp', 'Tmpo Lrtp'],
      dtype=object)

In [210]:
all_mpo_df.shape

(16276, 10)

In [211]:
all_mpo_df.data_source.value_counts()

Fresno Cog Lrtp    3147
Scag Lrtp          2952
Lost               1849
Sacog Lrtp         1601
Kern Cog Lrtp      1411
Scrtpa Lrtp        1066
Madera Ctc Lrtp     765
Stancog Lrtp        552
Slocog Lrtp         420
Sbcag Lrtp          419
Sandag Lrtp         416
Tcag Lrtp           337
Mtc Lrtp            282
Ambag Lrtp          280
Sjcog Lrtp          262
Bcag Lrtp           250
Mcagov Lrtp         108
Kcag Lrtp            84
Tmpo Lrtp            75
Name: data_source, dtype: int64

In [212]:
# all_mpo_df.drop(columns = ['notes']).sample(3)

In [213]:
gcs_test = pd.read_excel("gs://calitp-analytics-data/data-analyses/project_list/LRTP/all_LRTP_LOST.xlsx")

In [214]:
gcs_test.sample()

project_title lead_agency project_year    project_description  \
5668  Bus Shelters #3  Madera Ctc          NaN  New Bus Stop Shelters   

      total_project_cost  city county      data_source  \
5668           789078.00  None   None  Madera Ctc Lrtp   

                                         notes  
5668  Category: Public Transit,  Location: Nan

In [215]:
"""
all_mpo_gdf.explore(
    "project_title",
    cmap="tab20c",
    height=400,
    width=600,
    style_kwds={"weight": 4},
    legend=False,
)
"""

'\nall_mpo_gdf.explore(\n    "project_title",\n    cmap="tab20c",\n    height=400,\n    width=600,\n    style_kwds={"weight": 4},\n    legend=False,\n)\n'

### Cordon Pricing

In [216]:
congest_keywords = [
    "congestion pricing",
    "variable tolls",
    "express lane",
    "value pricing",
    "rush hour",
    "cordon",
    "dynamic pricing",
    "dynamically  priced",
    "high occupancy",
    "mobility pricing",
    "occupancy",
    "toll lane",
    "performance pricing",
    "peak travel",
    "managed lane",
    "tollway",
    "express toll",
    "fixed pricing",
    "hot lane",
    "hov lane",
    "expressed toll lane",
]

In [217]:
cordon_gdf, cordon_df = list_utils.filter_projects(
    all_mpo_df,
    [
        "project_title",
        "project_description",
    ],
    congest_keywords,
    "congestion_pricing_lrtp_lost",
    True,
)

In [218]:
len(cordon_df), len(cordon_gdf)

(175, 85)

In [219]:
len(cordon_df) / len(all_mpo_df)

0.010752027525190465

In [220]:
cordon_df.project_title.nunique(), cordon_gdf.project_title.nunique()

(98, 52)

In [221]:
cordon_df.groupby(["data_source"]).agg({"project_title": "nunique"})

project_title
data_source                   
Fresno Cog Lrtp              2
Lost                         8
Mtc Lrtp                     4
Sacog Lrtp                  30
Sandag Lrtp                 46
Sbcag Lrtp                   4
Scag Lrtp                    1
Sjcog Lrtp                   1
Slocog Lrtp                  1
Stancog Lrtp                 1
Tmpo Lrtp                    1

In [222]:
gdf_map_subset = [
    "project_title",
    "lead_agency",
    "project_year",
    "total_project_cost",
    "geometry",
]

In [223]:
"""
cordon_gdf[gdf_map_subset].explore(
    cmap="tab20c",
    height=400,
    width=600,
    style_kwds={"weight": 4},
    legend=False,
)
"""

'\ncordon_gdf[gdf_map_subset].explore(\n    cmap="tab20c",\n    height=400,\n    width=600,\n    style_kwds={"weight": 4},\n    legend=False,\n)\n'

### Categorizing

In [224]:
all_mpo_df = all_mpo_df.fillna("None")

In [225]:
categorized_df = categorizing_utils.add_categories(all_mpo_df)

In [226]:
drop_cols = [
    "lower_case_project_title",
    "lower_case_project_description",
    "lower_case_project_title_keyword_search",
    "lower_case_project_description_keyword_search",
    "geometry",
    "city",
    "county",
]

In [227]:
categorized_df = categorized_df.drop(columns=drop_cols)

In [228]:
categorized_df.n_categories.value_counts()

0.00    7351
1.00    5056
2.00    2605
3.00     877
4.00     297
5.00      71
6.00      19
Name: n_categories, dtype: int64

In [229]:
all_projects_categorized_by_mpo = categorizing_utils.find_categories_by_mpo(categorized_df)

In [230]:
all_projects_categorized_by_mpo.head()

data_source             category  total_projects
0  Ambag Lrtp        Active Transp          235.00
1  Ambag Lrtp               Bridge           10.00
2  Ambag Lrtp    Congestion Relief            0.00
3  Ambag Lrtp              Freeway            9.00
4  Ambag Lrtp  Infra Resiliency Er           39.00

In [231]:
categorizing_utils.count_all_categories(categorized_df)

Category  Total Projects
0         Active Transp         3140.00
1               Transit         1612.00
2                Bridge          769.00
3                Street         4737.00
4               Freeway          733.00
5   Infra Resiliency Er         2820.00
6     Congestion Relief          108.00
7  Passenger Mode Shift          111.00
8                Safety          524.00